subsys/debug/cpu_load/cpu_load.c

```diff
@@ -0,0 +1,234 @@
+		 * assuming that all subscribe registers has the same mask for
+		 * channel id.
+		 */
+		*ch = *PUBLISH_ADDR(evt) & DPPIC_SUBSCRIBE_CHG_EN_CHIDX_Msk;
+		err = NRFX_SUCCESS;
+	} else {
+		err = nrfx_dppi_channel_alloc(ch);
+	}
+#else
+	err = nrfx_ppi_channel_alloc((nrf_ppi_channel_t *)ch);
+#endif
+	return err;
+}
+
+static void cpu_load_log_fn(struct k_work *item)
+{
+	u32_t load = cpu_load_get();
+	u32_t percent = load/1000;
+	u32_t fraction = load - 1000*percent;
+
+	cpu_load_reset();
+	LOG_INF("Load:%d,%03d%%", percent, fraction);
+	k_delayed_work_submit(&cpu_load_log,
+			CONFIG_CPU_LOAD_LOG_INTERVAL);
+}
+
+static int cpu_load_log_init(void)
+{
+	k_delayed_work_init(&cpu_load_log, cpu_load_log_fn);
+	return k_delayed_work_submit(&cpu_load_log,
+			      CONFIG_CPU_LOAD_LOG_INTERVAL);
+}
+
+void timer_handler(nrf_timer_event_t event_type, void *context)
+{
+	/*empty*/
+}
+
+int cpu_load_init(void)
+{
+	u8_t ch_sleep;
+	u8_t ch_wakeup;
+	u8_t ch_tick = 0;
+	nrfx_err_t err;
+	nrfx_timer_config_t config = NRFX_TIMER_DEFAULT_CONFIG;
+	int ret = 0;
+
+	config.frequency = NRF_TIMER_FREQ_1MHz;
+	config.bit_width = NRF_TIMER_BIT_WIDTH_32;
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_ALIGNED_CLOCKS)) {
+		/* It's assumed that RTC1 is driving system clock. */
+		config.mode = NRF_TIMER_MODE_COUNTER;
+		err = ppi_alloc(&ch_tick,
+		       nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK));
+		if (err != NRFX_SUCCESS) {
+			return -ENODEV;
+		}
+		nrfx_gppi_channel_endpoints_setup(ch_tick,
+		     nrf_rtc_event_address_get(NRF_RTC1, NRF_RTC_EVENT_TICK),
+		     nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_COUNT));
+		nrf_rtc_event_enable(NRF_RTC1, NRF_RTC_INT_TICK_MASK);
+	}
+
+	if (ready) {
+		return 0;
+	}
+
+	err = nrfx_timer_init(&timer, &config, timer_handler);
+	if (err != NRFX_SUCCESS) {
+		return -EBUSY;
+	}
+
+	err = ppi_alloc(&ch_sleep,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPENTER));
+	if (err != NRFX_SUCCESS) {
+		return -ENODEV;
+	}
+
+	err = ppi_alloc(&ch_wakeup,
+		       nrf_power_event_address_get(NRF_POWER,
+						   NRF_POWER_EVENT_SLEEPEXIT));
+	if (err != NRFX_SUCCESS) {
+		/* ppi free */
+		return -ENODEV;
+	}
+
+	nrfx_gppi_channel_endpoints_setup(ch_sleep,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPENTER),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_START));
+	nrfx_gppi_channel_endpoints_setup(ch_wakeup,
+		  nrf_power_event_address_get(NRF_POWER,
+					      NRF_POWER_EVENT_SLEEPEXIT),
+		  nrfx_timer_task_address_get(&timer, NRF_TIMER_TASK_STOP));
+
+	nrfx_gppi_channels_enable((1 << ch_sleep) |
+				  (1 << ch_wakeup) |
+				  (1 << ch_tick));
+
+	cpu_load_reset();
+
+	if (IS_ENABLED(CONFIG_CPU_LOAD_LOG_PERIODIC)) {
+		ret = cpu_load_log_init();
+	}
+
+	ready = true;
+
+	return ret;
+}
+
+static u32_t cycle_ref;
+
+void cpu_load_reset(void)
+{
+	nrfx_timer_enable(&timer);
+	nrfx_timer_disable(&timer);
+	cycle_ref = k_cycle_get_32();
```

Comment:
Perhaps it would be good to use a a critical section for the above, so that there is no interruption between these operation? Similarly, for `nrfx_timer_capture()` with `k_cycle_get_32()` in `cpu_load_get()`.